### PR TITLE
feat(bulma-ui): add dialog-parity a11y behaviours and portal to Modal/Dialog

### LINKED ISSUE #617: [A11y] bulma-ui: Modal implements none of the dialog behaviours Dialog already has, and cannot portal
## Affected Package(s)

- [x] bulma-ui (`@allxsmith/bestax-bulma`)
- [ ] create-bestax (`create-bestax`)
- [x] docs (`@allxsmith/bestax-docs`)
- [ ] Other (please specify):

---

**Why this qualifies:** (a) defect for the dialog behaviours: the docs claim ARIA roles the component does not render, and `Dialog`, which wraps `Modal`, already implements every behaviour Modal lacks. (b) genuine deficiency for portalling: every other overlay in the library portals, and an inline `.modal` is clipped by any ancestor with `overflow: hidden`, `filter` or `transform`.

Surfaced while building the rbx migration source (#409, PR #613). rbx's modal closed on Escape, locked scroll, and rendered into `document.body` by default, so migrated apps lose all three silently.

## Summary

`Modal` is a positioned shell with a click-to-close backdrop and nothing else. `Dialog` implements Escape, initial focus, ref-counted scroll lock, `role` and `aria-modal`. The WAI-ARIA dialog pattern expects Escape to dismiss and the dialog to be announced as modal; Modal does neither, and its docs say it does. Separately, neither component can portal, while Toast, Sidebar, Notification and the picker popover all do.

### Dialog behaviours

`bulma-ui/src/components/Modal.tsx` (431 lines):
- Zero matches for `keydown`, `Escape`, `useEffect`, `role=`, `aria-modal`, `is-clipped`, `overflow`. The only ARIA is `aria-label="close"` on the close buttons at `:268`, `:377`, `:411`.
- Legacy path: `.modal-background` with `onClick={onClose}` at `:356-362`. Compound path renders children verbatim at `:341-348`; `Modal.Background` (`:122-131`) has no default `onClick`, so even backdrop dismissal is caller-supplied there.
- Root is `ModalRoot: React.FC<ModalProps>` at `:300`.

Docs overstate, then punt: `docs/docs/api/components/modal.md:278` "uses Bulma's structure and ARIA roles for accessibility" (no role is rendered); `:281` "for focus trap or escape key handling, implement those patterns in your app as needed."

`bulma-ui/src/components/Dialog.tsx` (which renders `<Modal>` at `:265`) already does all of it: Escape `:123-135`, focus management `:137-144`, ref-counted body scroll lock `:146-162`, `role="alertdialog"` + `aria-modal="true"` `:270-271`, documented at `docs/docs/api/components/dialog.md:321-325`.

Bulma ships the scroll-lock helper (`bulma/sass/helpers/overflow.scss:3`, `.is-clipped`) and bestax exposes it as `overflow="clipped"` (`bulma-ui/src/helpers/useOtherClasses.tsx:12-13`), but Modal never applies it to `<html>`.

### Portal

- `grep -rn createPortal` over `Modal.tsx` and `Dialog.tsx` returns nothing. `position: fixed` does not escape a containing block created by `transform`, `filter`, `perspective` or `contain: paint`, and an `overflow: hidden` ancestor clips the backdrop.
- Established pattern in the same package: `Toast.tsx:8,296,472`, `Sidebar.tsx:2,221`, `elements/Notification.tsx:2,388`, `form/_pickerInternals/PickerPopover.tsx:8,190`.
- Two prop shapes to mirror: Sidebar is portal-by-default with an `inline` opt-out (`Sidebar.tsx:41`); Toast takes `container?: string | HTMLElement` and resolves it at `Toast.tsx:65, 208-215`.

### Parity behind the existing API (no shape change; #500 recorded that nothing supports changing the Modal/Dialog shape)

- `closeOnEscape?: boolean` (default `true`): `keydown` listener while active, calls `onClose`.
- `lockScroll?: boolean` (default `true`): lift Dialog's ref-counted lock into a shared helper so Dialog-inside-Modal does not double count, or toggle `is-clipped` on `document.documentElement`.
- `role="dialog"` and `aria-modal="true"` on the root when active; pass through a caller `role` so Dialog's `alertdialog` still wins.
- Move focus into the modal on open and restore it on close.
- Apply to both the legacy and compound render paths.

### Opt-in portal

- `portal?: boolean | string | HTMLElement` on `ModalProps` (default `false`, so nothing changes for current users). `true` targets `document.body`; a string is a selector; an element is used directly. Reuse Toast's resolver by lifting it into a shared helper. `Dialog` forwards the prop. SSR guard: render inline when `document` is undefined, matching `Dropdown`'s `isBrowser` at `Dropdown.tsx:12`.
- Flipping the default to `true` is a `next-major` candidate; note it in the TSDoc.

## Non-goals

No Tab-cycle focus trap in this issue (`aria-modal="true"` already marks the rest of the page inert for AT; a trap can follow). No change to the portal default. No z-index management. No new sub-components. No changes to Toast/Sidebar/Notification beyond extracting the shared helpers.

## Acceptance

- [ ] Escape, scroll lock, `role`/`aria-modal`, and focus move/restore implemented on both render paths, each behind a default-on prop
- [ ] `portal` prop on `Modal` and `Dialog`, default `false`, with `true` / selector / element and the inline SSR fallback covered
- [ ] Tests in `bulma-ui/src/components/__tests__/Modal.test.tsx` for each behaviour, the opt-outs, and each portal form; `Dialog.test.tsx` still green with no double scroll lock
- [ ] `pnpm typecheck` green
- [ ] `pnpm test` green, bulma-ui coverage >= 99%
- [ ] `pnpm check:conformance` green
- [ ] `pnpm gen` run for the props tables
- [ ] Story + docs page updated (`Modal.stories.tsx` gains a portal story; `modal.md:278-281` rewritten to state what actually ships; `dialog.md` documents `portal`), since CI only checks they exist

### PR DESCRIPTION
**Pull request by bestaxbot**
## Summary

`Modal` was a positioned shell with only a click-to-close backdrop — no Escape handling, no scroll lock, no ARIA role/`aria-modal`, no focus management — despite its docs claiming ARIA support, and `Dialog` (which wraps `Modal`) already implemented all of it independently. Separately, neither component could portal, unlike every other overlay in the library (`Toast`, `Sidebar`, `Notification`, the picker popover).

- `Modal` gains `closeOnEscape` (default `true`), `lockScroll` (default `true`), and `portal` (default `false`) props. It now applies `role="dialog"`/`aria-modal="true"` while active (a caller-supplied `role` wins — e.g. `presentation` skips `aria-modal` too), and moves focus into the modal on open and restores it on close. Both the legacy and compound render paths get all of this.
- Body scroll locking is now a shared, ref-counted `useScrollLock` helper (`bulma-ui/src/helpers/scrollLock.ts`) so a `Dialog` (which renders its own `Modal`) doesn't double up with another open overlay.
- Portal target resolution is a shared `resolvePortalContainer` helper (`bulma-ui/src/helpers/portal.ts`), also adopted by `Toast` (previously had its own copy of the same logic). `portal` accepts `true` (portals to `document.body`), a `querySelector` string, or an `HTMLElement`, and is ignored (renders inline) when `document` is undefined so it's SSR-safe.
- `Dialog` now delegates Escape and scroll-lock entirely to the `Modal` it wraps (passing `onClose`) instead of duplicating that logic, and forwards a new `portal` prop straight through. It passes `role="presentation"` to the wrapping `Modal` so its own `role="alertdialog"` on the inner content div remains the only one announced (no duplicate dialog semantics).

## Test plan

- [x] `pnpm --filter @allxsmith/bestax-bulma test:coverage` — 100% on `Modal.tsx`, `Dialog.tsx`, `Toast.tsx`, and the two new helpers; overall coverage above the 99% gate
- [x] New tests: `Modal.test.tsx` (Escape, scroll lock, role/aria-modal, focus move/restore, portal — selector/element/boolean forms), `Modal.ssr.test.tsx` (portal ignored when `document` is undefined), `Dialog.test.tsx` (single alertdialog role, portal forwarding), `helpers/__tests__/{scrollLock,portal}.test.ts`
- [x] `pnpm --filter @allxsmith/bestax-bulma run lint` / `typecheck` / `build` / `build-storybook`
- [x] `pnpm run format:check` and `pnpm run check:conformance`
- [x] `pnpm gen` run — regenerated `docs/docs/api/components/{modal,dialog}.md` props tables and `bestax-mcp/data/**`
- [x] New Storybook story `Modal.stories.tsx` `Portal` export; `modal.md`/`dialog.md` prose (Usage + Accessibility) rewritten to describe what actually ships
- [x] `pnpm test` (full monorepo) — all suites green

## Known follow-up (out of scope here)

`skills/bestax-migrate/references/rbx/unmappables.md` documents the old gap (rbx Modal's Escape/scroll-lock/portal defaults not carrying over) and the codemod emits a matching TODO for every migrated `Modal`. That guidance is now stale since these are Modal defaults, but updating the codemod's TODO logic and its tests is a separate, sizable change in `bestax-migrate` — filing as a follow-up rather than bundling it here.

Fixes #617

Generated with [Claude Code](https://claude.ai/code)

## Summary by CodeRabbit

* **New Features**
  * Added portal rendering for Dialog and Modal, supporting custom targets.
  * Added configurable Escape-key closing and body scroll locking for Modal.
  * Improved modal focus management and accessibility behavior.
  * Dialog now supports portal rendering and forwards relevant modal behavior.

* **Documentation**
  * Updated Dialog and Modal guides with portal, accessibility, focus, and keyboard behavior details.

* **Bug Fixes**
  * Improved server-side rendering compatibility for portaled modals.
  * Prevented duplicate dialog accessibility roles.

### CODE REVIEW

File: bulma-ui/src/components/Modal.tsx [hunk 1]
```diff
@@ -320,6 +349,41 @@ const ModalRoot: React.FC<ModalProps> = ({
...
+  // Move focus into the modal on open, restore it on close
+  useEffect(() => {
+    if (!isModalActive) return undefined;
+
+    previouslyFocusedRef.current = document.activeElement as HTMLElement | null;
+    const node = modalRootRef.current;
+    const focusable = node?.querySelector<HTMLElement>(
+      'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
+    );
+    (focusable ?? node)?.focus();
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx]: Good catch — this is a real bug in the focus-management I added. A disabled first control matched the selector, and calling `focus()` on a disabled element is a no-op, so focus would stay outside the modal whenever a disabled control preceded an enabled one.

Fixed in c0c34166a7736d501180696807dc6192859a454d — I narrowed the `querySelector` in the focus effect (`Modal.tsx:377`) to `button:not(:disabled), [href], input:not(:disabled):not([type="hidden"]), select:not(:disabled), textarea:not(:disabled), [tabindex]:not([tabindex="-1"])`, and added a regression test ("skips a disabled first control and focuses the first enabled one") that renders a disabled button before an enabled one and asserts focus lands on the enabled control.

File: bulma-ui/src/components/Modal.stories.tsx [hunk 2]
```diff
@@ -156,3 +156,29 @@ const CompoundModalContentComponent = () => {
 export const CompoundModalContent: StoryObj<ModalProps> = {
   render: () => <CompoundModalContentComponent />,
 };
+
+// Portal — renders into document.body instead of inline
+const PortalModalComponent = () => {
+  const [open, setOpen] = useState(true);
+  return (
+    <div style={{ overflow: 'hidden', height: 120, border: '1px dashed' }}>
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.stories.tsx]: Fixed in 46915f12ce — this was already addressed before this fix pass. The Portal story's clipping wrapper no longer uses a 4th inline `style={{}}`; it's now a `Box` with `is-clipped` (Bulma's `overflow: hidden` helper) plus `m`/`p` helper props, so `Modal.stories.tsx` is back to 3 inline styles and the baseline stays at 3. `node scripts/check-conformance.mjs --only=inline-style` passes green now, so nothing in `conformance-baseline.json` needed to move.

**Comment by allxsmith** [comment]
@bestaxbot please address the open review threads on this PR. For each finding: verify it against the current head before changing anything, fix the ones that hold and add a test for each, reply with the reason on any you decline, and resolve every thread you have handled. Keep the change within this PR's scope, and if the branch conflicts with `main`, merge `main` in first.

Note Copilot reviews do not wake you on their own (its author_association is not OWNER/MEMBER/COLLABORATOR), so treat Copilot threads as in scope here alongside CodeRabbit's. Also read any suppressed comments in the review bodies, not just the inline threads.

File: bulma-ui/src/components/Modal.tsx [hunk 3]
```diff
@@ -320,6 +349,41 @@ const ModalRoot: React.FC<ModalProps> = ({
...
+    if (!isModalActive || !closeOnEscape) return undefined;
+
+    const handleKeyDown = (e: KeyboardEvent) => {
+      if (e.key === 'Escape') {
+        onClose?.();
+      }
+    };
+
+    document.addEventListener('keydown', handleKeyDown);
+    return () => document.removeEventListener('keydown', handleKeyDown);
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx:453]: Fixed in `a05c5b14`. You're right, and it's a real bug: every active Modal installed its own `document` keydown listener, so one Escape fanned out to all of them — and the ref-counted scroll lock this PR added does explicitly contemplate overlapping overlays, so "two open at once" is a supported state.

There is now a module-level `activeModalStack` (`Modal.tsx:23`). Each active modal pushes an identity token in an effect keyed on `isModalActive` and splices it out on cleanup (`Modal.tsx:392-400`), and the keydown handler returns early unless its own token is on top (`Modal.tsx:407-412`). Escape and the new Tab containment both go through that one check. Registration is keyed on `isModalActive` rather than `closeOnEscape`, so a topmost modal with `closeOnEscape={false}` swallows Escape instead of letting it fall through to the modal underneath.

Tests: `closes only the topmost modal when several are open` and `closes the next modal down once the topmost one has closed`.

File: bulma-ui/src/components/Modal.tsx [hunk 4]
```diff
@@ -320,6 +349,41 @@ const ModalRoot: React.FC<ModalProps> = ({
...
+
+    previouslyFocusedRef.current = document.activeElement as HTMLElement | null;
+    const node = modalRootRef.current;
+    const focusable = node?.querySelector<HTMLElement>(
+      'button:not(:disabled), [href], input:not(:disabled):not([type="hidden"]), select:not(:disabled), textarea:not(:disabled), [tabindex]:not([tabindex="-1"])'
+    );
+    (focusable ?? node)?.focus();
+
+    return () => {
+      previouslyFocusedRef.current?.focus?.();
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx:477]: Fixed in `a05c5b14`. The restore is now conditional (`Modal.tsx:460-470`): it bails out when `document.activeElement` is a real element that is not inside this modal — exactly the background-close case you describe. Focus sitting on `<body>` still counts as ours, because a modal whose subtree is removed leaves focus there and that path must still restore; the existing `restores focus to the previously focused element on close` test covers it.

I went with a containment check rather than stack ownership on purpose. Ownership would have to be read either side of the stack effect's cleanup, so it depends on the relative order of two effect cleanups in the same component — subtle, and easy to break later. "Do I still hold focus?" is order-independent and additionally covers the case where the user clicked something outside before the modal closed, which stack ownership would not.

Test: `does not pull focus out of a modal that is still open on top` — two open modals, close the background one, assert focus stays on the top modal's control and does not return to the element saved before it opened.

File: bulma-ui/src/components/Modal.tsx [hunk 5]
```diff
@@ -338,84 +402,123 @@ const ModalRoot: React.FC<ModalProps> = ({
 
   const modalClasses = classNames(bulmaClasses, bulmaHelperClasses, className);
 
+  const resolvedRole = role ?? (isModalActive ? 'dialog' : undefined);
+  const resolvedAriaModal =
+    ariaModalProp ??
+    (isModalActive && resolvedRole !== 'presentation' ? 'true' : undefined);
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx:507]: Half fixed, half documented — taking the first of your two options for the keyboard half.

**Fixed:** Tab and Shift+Tab now cycle inside the topmost modal (`Modal.tsx:414-443`) — wrap forward at the last control, backward at the first, pull focus back in if it is outside the modal, and fall back to focusing the modal root when nothing inside is focusable. Five tests cover those paths plus the mid-list no-op (we must not hijack Tab between the first and last control).

**Not fixed, documented instead:** I did not apply `inert` to background content. A Modal rendered inline sits inside the tree we would have to inert, so "background" is every sibling subtree on the path to the root — not something the component can identify without owning the whole document. That's why the mainstream React modals ship `aria-modal` plus Tab containment and leave `inert` to the app. So `modal.md` now says plainly that `aria-modal` does not by itself make background content inert, that no `inert` attribute is applied, and that background content stays reachable by pointer and by browser UI such as find-in-page.

I did not take your second option (leave `aria-modal` unset) because it would leave Modal with no modal semantics at all, which is the gap #617 was opened about.

File: bulma-ui/src/components/Modal.tsx [hunk 6]
```diff
@@ -338,84 +402,123 @@ const ModalRoot: React.FC<ModalProps> = ({
...
   // If using compound components, render children as-is
   if (hasCompoundComponents) {
-    return (
-      <div className={modalClasses} {...rest} data-testid="modal">
+    modalElement = (
+      <div
+        className={modalClasses}
+        ref={modalRootRef}
+        role={resolvedRole}
+        aria-modal={resolvedAriaModal}
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx:522]: Fixed in `a05c5b14` for the legacy path, documented for the compound one.

`modalCardTitle` now gets a `useId`-generated id on the title element (`Modal.tsx:500`, `Modal.tsx:549`) and the modal root carries `aria-labelledby` pointing at it (`Modal.tsx:532`). A caller-supplied `aria-labelledby` still wins, because it arrives through `rest`, which is spread after — there's a test asserting that.

For the compound API the component genuinely cannot pick the title: `Modal.Card.Title` is optional, and the content may be an arbitrary `Modal.Content` tree. Generating an id for a node we did not author would be guessing. So instead `modal.md`'s Accessibility section now leads with "give the dialog an accessible name" and states that compound users pass their own `aria-label`/`aria-labelledby` on `<Modal>`.

Three tests: title wiring (resolves the id and asserts the target's text), no `aria-labelledby` when there's no card title, and caller override.

**Comment by allxsmith** [comment]
@bestaxbot your last pass (`a05c5b1`) answered four threads and the session ended with four still untouched. CI is green on that head. Copilot reviews do not wake you, so this is the trigger.

The four outstanding, all Copilot, and none of them is a nitpick:

1. **SSR hydration** — `portal` branches on `document` during render, so the server emits inline markup and the client emits a portal. That is a hydration mismatch, not a styling detail. Decide it deliberately: render inline until mounted, or document that `portal` is client-only.
2. **Scroll-lock coordination (raised twice, at 02:37 and 04:01 — treat as one)** — the new ref count only coordinates callers of the hook, while `Sidebar.tsx:124-131` and `Loading.tsx:143-149` still set `document.body.style.overflow` directly. A Sidebar closing under an open Modal can unlock the page. Either route those two through the same hook or say why they are out of scope for this PR.
3. **`aria-modal` does not make background content inert** — it announces modal semantics only. If the PR claims inertness anywhere in docs or TSDoc, the claim needs narrowing to what the attribute actually delivers.

Same rules: verify each against the current head first, fix what holds with a test, and reply with your reasoning on any you decline rather than leaving it silent. If a finding is real but genuinely belongs in its own PR, say that explicitly and I will open one.

File: bulma-ui/src/components/Modal.tsx [hunk 7]
```diff
@@ -338,84 +402,123 @@ const ModalRoot: React.FC<ModalProps> = ({
...
+        )}
+      </div>
+    );
+  }
+
+  if (portal && typeof document !== 'undefined') {
+    const target = resolvePortalContainer(
+      typeof portal === 'boolean' ? undefined : portal
+    );
+    return createPortal(modalElement, target);
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx]: You were right, and this is fixed in `a05c5b14` — I'd pushed it before the last session ended without replying here, so the thread stayed silent.

The render no longer branches on `document`. `Modal.tsx:382-386` reads an "are we past hydration?" flag through `useSyncExternalStore` with a server snapshot of `false` and a client snapshot of `true` (`Modal.tsx:33-37`), and the portal branch is now `if (portal && onClient)` (`Modal.tsx:608`). React uses the server snapshot both when server-rendering *and* while hydrating, so the first client render is inline and identical to the server's; the store's snapshot change then commits a re-render that moves the tree into the portal. That's the deliberate decision you asked for: portal is a post-mount transition, not a client-only prop.

Regression test as requested — `hydrates the server markup inline, then moves into the portal` (`Modal.test.tsx:801`). It `renderToString`s the modal, hydrates that markup with `hydrateRoot`, asserts `console.error` was never called (a hydration mismatch logs there), and then asserts the modal has left the container and its `parentElement` is `document.body`. The node-environment `Modal.ssr.test.tsx` still covers the server half, where `document` is genuinely undefined.

File: bulma-ui/src/helpers/scrollLock.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * @param active - Whether this caller wants the body scroll locked.
+ */
+export function useScrollLock(active: boolean): void {
+  useEffect(() => {
+    if (!active) return undefined;
+
+    lockCount++;
+    if (lockCount === 1) {
+      originalOverflow = document.body.style.overflow;
+      document.body.style.overflow = 'hidden';
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/helpers/scrollLock.ts:26]: Fixed in `a05c5b14` (pushed before the last session ended, but I never replied here — sorry).

`Sidebar` and `Loading` no longer touch `document.body.style.overflow` themselves; they route through the same hook (`Sidebar.tsx:127`, `Loading.tsx:147`). `grep -rn 'body.style.overflow' bulma-ui/src` outside tests now returns only `scrollLock.ts` itself, so the ref count coordinates every body-locking overlay in the library rather than just Modal/Dialog.

Cross-component overlap tests, both close orders you asked for — `stays locked when an overlapping Loading closes first` and `stays locked when an overlapping Sidebar closes first` (`Modal.test.tsx:595` and `:610`). Each renders a Modal alongside the other overlay, closes the other one, and asserts `overflow` is still `hidden` while the Modal is open. The pre-existing Sidebar and Loading tests that assert their own lock/restore behaviour still pass unchanged, so routing through the shared hook didn't regress the single-overlay case.

File: bulma-ui/src/components/Modal.tsx [hunk 9]
```diff
@@ -320,6 +349,41 @@ const ModalRoot: React.FC<ModalProps> = ({
   // Support both active and isActive props
   const isModalActive = active ?? isActive ?? false;
 
+  const modalRootRef = useRef<HTMLDivElement>(null);
+  const previouslyFocusedRef = useRef<HTMLElement | null>(null);
+
+  useScrollLock(isModalActive && lockScroll);
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx:396]: Same finding as the one on `scrollLock.ts:26`, and it's fixed in `a05c5b14` — I've written the detail there, but repeating the short version so this thread isn't left silent.

`Sidebar` and `Loading` now call `useScrollLock` (`Sidebar.tsx:127`, `Loading.tsx:147`) instead of setting `document.body.style.overflow` directly, so every body-scroll lock in the library shares one ref count. Both mixed-overlay close orders are covered: `stays locked when an overlapping Loading closes first` (`Modal.test.tsx:595`) and `stays locked when an overlapping Sidebar closes first` (`:610`).

File: docs/docs/api/components/modal.md [hunk 10]
```diff
@@ -275,10 +306,13 @@ function example() {
...
 
-- The modal root uses Bulma’s structure and ARIA roles for accessibility.
+- The modal root gets `role="dialog"` and `aria-modal="true"` while active — pass an explicit `role` to override (e.g. `Dialog` passes `role="presentation"` on the `Modal` it wraps so its own `role="alertdialog"` is the only one announced).
+- Escape closes the modal (calls `onClose`) by default; opt out with `closeOnEscape={false}`.
+- Body scroll is locked while the modal is active (ref-counted, so a `Dialog` rendering its own `Modal` doesn't fight another open modal for the unlock); opt out with `lockScroll={false}`.
+- Focus moves to the first focusable element (or the modal itself) on open, and is restored to the previously focused element on close.
 - The modal background closes the modal on click (`onClose` required).
 - Close buttons are provided for both modal-card and modal-content variants.
-- Keyboard/screen reader accessibility is supported, but for focus trap or escape key handling, implement those patterns in your app as needed.
+- No Tab-cycle focus trap is implemented — `aria-modal="true"` already marks the rest of the page inert for assistive technology.
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply docs/docs/api/components/modal.md]: Agreed on the substance, and the sentence is already gone as of `a05c5b14` — I fixed it alongside the sibling thread on `Modal.tsx` and didn't reply here.

Two changes. The keyboard half is now implemented rather than claimed: Tab and Shift+Tab cycle inside the topmost modal (`Modal.tsx:414-443`), wrapping at the last and first control, pulling focus back in when it's outside, and falling back to the modal root when nothing inside is focusable. And the docs no longer overstate the attribute — `modal.md:316` now reads that `aria-modal` does not by itself make background content inert, that no `inert` attribute is applied, and that background content stays reachable by pointer and by browser UI such as find-in-page. `dialog.md:347` carries the same narrowed wording.

I did not apply `inert` to background content, and that's deliberate rather than an omission. A Modal rendered inline lives inside the very tree we'd have to inert, so "background" is every sibling subtree along the path to the root — the component can't identify that without owning the document. Tab containment plus honest docs is where the mainstream React modals land too; `inert` stays the app's call.

**Comment by allxsmith** [comment]
@bestaxbot Copilot posted four more threads at 12:55, after your last pass, and none is answered. Copilot does not wake you, so this is the trigger. CI is green on `a05c5b1`.

Three of these look substantive, and one is a behaviour regression on a component this PR was not meant to change:

1. **`portal.ts` changes existing Toast behaviour** — for `container=""` the old resolver treated the falsy value as "no container" and used `document.body`, whereas `document.querySelector("")` throws. This PR extracted that resolver, so it is on the hook for keeping Toast working. Check the Toast tests, and preserve the falsy-means-body case unless you can show nothing relies on it.
2. **Focus-trap selector is not the tabbable set** — `<a href tabIndex={-1}>` still matches `[href]`, so `last` can point at something the browser will not tab to. Either filter on `tabIndex >= 0` or say why the simpler selector is acceptable here.
3. **Hydration remount drops the focus target** — switching from host element to portal remounts the subtree while the focus effect is keyed only on `isModalActive`, so it retains the pre-portal node. This is the flip side of the SSR fix you already made; key the effect on the portal target too.
4. **`Dialog.portal` has docs but no Storybook update** — same review-only gate as the story findings on #636, which turned out to be correct there.

Usual rules: verify each against the current head first, fix what holds with a test, and reply with reasoning on anything you decline rather than leaving it silent. If (1) is genuinely out of scope for this PR, say so and I will open a separate one for the Toast resolver.

File: bulma-ui/src/helpers/portal.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+ * @param container - The requested portal target, if any.
+ * @returns The resolved DOM node to portal into.
+ */
+export function resolvePortalContainer(
+  container?: string | HTMLElement
+): HTMLElement {
+  if (container && typeof container !== 'string') {
+    return container;
+  }
+  if (typeof container === 'string') {
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/helpers/portal.ts:21]: Confirmed and fixed in `a983085e`. You're right that this is a behaviour change on Toast, not a new-code detail.

The old resolver was `if (container) { … }` (`Toast.tsx:209` pre-extraction), so `""` fell out of the truthiness check and landed on `document.body`. My extracted version guarded with `container && typeof container !== 'string'` first and then re-tested `typeof container === 'string'` — which `""` passes, so it reached `document.querySelector('')` and threw a `SyntaxError`.

`portal.ts` now leads with `if (!container) return document.body;`, so any falsy value — `undefined` or `""` — short-circuits before the selector, and the element/selector branches follow. I checked the Toast tests first as you asked: nothing covered `container=""`, which is why this got through, so there are now two tests — `treats an empty string as no container rather than querying it` (`portal.test.ts`) and `falls back to document.body for an empty string container` (`Toast.test.tsx`), both asserting it doesn't throw and lands on `document.body`.

File: bulma-ui/src/components/Modal.tsx [hunk 12]
```diff
@@ -11,6 +12,29 @@ import {
...
+ */
+const activeModalStack: object[] = [];
+
+/**
+ * Controls the modal can hand focus to. Disabled and hidden controls are
+ * excluded because `focus()` on them is a no-op, which would leave focus
+ * outside the modal.
+ */
+const FOCUSABLE_SELECTOR =
+  'button:not(:disabled), [href], input:not(:disabled):not([type="hidden"]), select:not(:disabled), textarea:not(:disabled), [tabindex]:not([tabindex="-1"])';
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx]: Fixed the core of this in `a983085e`, and I'm declining the wider ask — reasoning below.

**Fixed.** Your `<a href tabIndex={-1}>` example is a real escape hatch: `[href]` and `button` match regardless of `tabindex`, so that anchor sorted last, `last` pointed at it, and Tab from the actual last control never hit the wrap branch. There's now a `getTabbable()` (`Modal.tsx:36`) that runs the selector and then filters on `el.tabIndex >= 0`. Both call sites go through it — the Tab handler and the initial-focus effect — so the two are guaranteed to agree, which the two separate selectors would not have been.

Filtering on the resolved `el.tabIndex` rather than widening the selector also covers your "negative tab indexes other than `-1`" point for free: the browser resolves `tabindex="-2"` to `-2`, so it drops out without a `:not([tabindex="-2"])`-style selector to keep in sync. I dropped the now-redundant `:not([tabindex="-1"])` from the selector for the same reason.

Three regression tests: `ignores a negative-tabIndex anchor when picking the last tab stop` (asserts both Tab and Shift+Tab), `ignores tab indexes below -1 when picking tab stops`, and `skips a negative-tabIndex first control when moving focus in` for the initial-focus path. I verified all three fail against the old unfiltered set before the fix.

**Declining: a full tabbable calculation.** CSS-hidden controls, radio-group roving, and `contentEditable` are all genuinely part of a correct tabbable set, but building one here is the wrong call:

- Visibility is the load-bearing half of it and is **untestable in this suite**. jsdom does no layout — `offsetParent` is always `null`, `getBoundingClientRect()` is all zeros, and `getComputedStyle` doesn't cascade. So a `display: none` filter would be code I can't cover, in a package with a 99% coverage gate, and any test I wrote would assert the mock rather than the behaviour.
- A correct implementation is the `tabbable` package, not a helper. Its shipped edge-case list is long enough that hand-rolling a partial copy in `Modal.tsx` is likelier to produce confidently-wrong containment than the current honest approximation. Adding a runtime dependency to a component library whose bundle size is marketing-visible (the READMEs carry a live bundlephobia badge) isn't something I'll do off a review comment.

So the line I've drawn is: fix what `tabIndex` alone decides, since that's exact, cheap, and testable; leave layout-dependent visibility to the app. If you want real tabbable semantics I'd rather that be its own PR with the dependency question decided deliberately — say the word and I'll write up the issue.

File: bulma-ui/src/components/Modal.tsx [hunk 13]
```diff
@@ -338,84 +490,129 @@ const ModalRoot: React.FC<ModalProps> = ({
...
+        )}
+      </div>
+    );
+  }
+
+  if (portal && onClient) {
+    const target = resolvePortalContainer(
+      typeof portal === 'boolean' ? undefined : portal
+    );
+    return createPortal(modalElement, target);
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Modal.tsx]: Confirmed on both counts, including the Dialog half you flagged at the end. Fixed in `a983085e`.

You traced it exactly: the effect was keyed `[isModalActive]`, so the inline→portal remount didn't re-run it. Focus stayed on the detached pre-portal node (falling to `<body>`), and on close the cleanup's captured `node` was the stale one — `node.contains(activeElement)` was false against a non-body `activeElement`, so the ownership guard I added for the overlapping-modal case took the early return and skipped restoration outright.

The render now computes `isPortaled = Boolean(portal) && onClient` (`Modal.tsx:387`), used both as the portal condition and as a dep on the focus effect. The ordering works out cleanly: React mutates the DOM before running passive effects, so by cleanup time the old subtree is detached and focus is on `<body>` — which the guard already treats as "still ours" — so cleanup restores to the pre-open element, and the re-run then re-records that same element before focusing into the portaled node. Restore-on-close still lands in the right place rather than being knocked off by the transition. For a non-portal modal `isPortaled` is constant `false`, so nothing re-runs.

Dialog had the identical bug for the same reason (`[isOpen, focusCancel, showCancel]`), so it now computes the same flag and adds it to its deps. To avoid a second copy of the `useSyncExternalStore` probe I moved it out of `Modal.tsx` into a shared `useIsHydrated()` (`helpers/useIsHydrated.ts`) that both import.

Tests, as you asked — focus after the move *and* after close. `keeps focus in the modal across the hydration portal move, and restores it on close` (`Modal.test.tsx`) hydrates a portaled modal with a button inside, asserts focus is on the portaled button, then unmounts and asserts focus returns to the element focused before open. `focuses the confirm button after the hydration portal move` (`Dialog.test.tsx`) covers the Dialog path. Both fail against the old deps — I checked.

File: bulma-ui/src/components/Dialog.tsx [hunk 14]
```diff
@@ -43,6 +39,12 @@ export interface DialogProps
   focusCancel?: boolean;
   /** Custom icon to display. */
   icon?: React.ReactNode;
+  /**
+   * Renders the dialog into a portal target instead of inline. Forwarded to
+   * the underlying `Modal`; see its `portal` prop for the accepted values.
+   * @defaultValue false
+   */
+  portal?: boolean | string | HTMLElement;
```
[the remark this answers is not in the document]
Comment (bestaxbot) [reply bulma-ui/src/components/Dialog.tsx:48]: Fair — the anatomy rule applies to `Dialog` here and I only added the story on the `Modal` side. Fixed in `a983085e`.

`Dialog.stories.tsx` now has a `Portal` export mirroring `Modal`'s: a `Box` with `is-clipped` as the clipping ancestor, and a `portal` Dialog that visibly escapes it. Using Bulma's `is-clipped` rather than an inline `overflow: hidden` also keeps it off the inline-style conformance ratchet, same as the Modal story. `build-storybook` and `check:conformance` both pass.

The prop already had its API docs entry, so this closes the story half of the rule.

**Review by allxsmith — APPROVED** [review]
